src/compartment.c

```diff
@@ -1108,26 +939,41 @@ seek_lib_data(void *lib_data, off_t offset)
     return (void *) ((char *) lib_data + offset);
 }
 
+void *
+get_seg_target(void *base_addr, struct LibDependency *lib_dep, size_t seg_id)
+{
+    return (char *) base_addr + (uintptr_t) lib_dep->lib_mem_base
+        + (uintptr_t) lib_dep->lib_segs[seg_id].mem_bot;
+}
+
 static void *
-eval_sym_offset(struct Compartment *comp, const comp_symbol *sym)
+eval_staged_sym_offset(struct Compartment *comp, const comp_symbol *sym)
 {
+    assert(comp->staged_addr != NULL);
     return (char *) comp->libs[sym->sym_lib_idx]->lib_mem_base
-        + (intptr_t) sym->sym_ref->sym_offset;
+        + (uintptr_t) sym->sym_ref->sym_offset;
 }
 
 static void *
-eval_lib_sym_offset(
-    struct Compartment *comp, const size_t lib_idx, const lib_symbol *sym)
+eval_staged_sym_tls_offset(struct Compartment *comp, const comp_symbol *sym)
 {
-    return (char *) comp->libs[lib_idx]->lib_mem_base
-        + (intptr_t) sym->sym_offset;
+    return (char *) sym->sym_ref->sym_offset
+        + comp->libs[sym->sym_lib_idx]->tls_offset; // TODO check offset type
 }
 
 static void *
-eval_sym_tls_offset(struct Compartment *comp, const comp_symbol *sym)
+eval_sym_offset(struct Compartment *comp, const comp_symbol *sym)
 {
-    return (char *) sym->sym_ref->sym_offset
-        + comp->libs[sym->sym_lib_idx]->tls_offset;
+    return (char *) comp->libs[sym->sym_lib_idx]->lib_mem_base
+        + (uintptr_t) sym->sym_ref->sym_offset;
+}
+
+static void *
+eval_lib_sym_offset(
+    struct Compartment *comp, const size_t lib_idx, const lib_symbol *sym)
+{
+    return (char *) comp->libs[lib_idx]->lib_mem_base
+        + (uintptr_t) sym->sym_offset;
```

Comment:
What type is `sym_offset`? Depending on that, the cast to `uintptr_t` might be dangerous.

Reply:
`sym_offset` is `void*` [1], so this cast seems appropriate.

[1] https://github.com/capablevms/CHERI-ELF-comp/blob/master/include/symbols_lib.h#L14